runtime/src/main/cpp/Memory.h

```diff
@@ -142,11 +142,14 @@ OBJ_GETTER(AllocInstance, const TypeInfo* type_info) RUNTIME_NOTHROW;
 
 OBJ_GETTER(AllocArrayInstance, const TypeInfo* type_info, int32_t elements);
 
-OBJ_GETTER(InitInstance,
-    ObjHeader** location, const TypeInfo* typeInfo, void (*ctor)(ObjHeader*));
+OBJ_GETTER(InitThreadLocalSingleton, ObjHeader** location, const TypeInfo* typeInfo, void (*ctor)(ObjHeader*));
 
-OBJ_GETTER(InitSharedInstance,
-    ObjHeader** location, const TypeInfo* typeInfo, void (*ctor)(ObjHeader*));
+OBJ_GETTER(InitSingleton, ObjHeader** location, const TypeInfo* typeInfo, void (*ctor)(ObjHeader*));
+
+// `initialValue` may be `nullptr`, which signifies that the appropriate initial value was already
```

Comment:
Note: for statically-initialized read-only globals it might make sense to avoid registering at all, because the value is not going to be garbage collected anyway.
This optimization is not required at the moment, but we should at least have this machinery flexible enough.

Reply:
It was relatively easy to add, so done.